hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapReplicationStateHolder.java

```diff
@@ -162,12 +165,30 @@ void applyState() {
                     indexes.addOrGetIndex(indexDefinition.getKey(), indexDefinition.getValue());
                 }
 
+                final Indexes indexes = mapContainer.getIndexes(partitionContainer.getPartitionId());
+                final SerializationService serializationService = mapContainer.getMapServiceContext().getNodeEngine()
+                                                                              .getSerializationService();
                 for (RecordReplicationInfo recordReplicationInfo : recordReplicationInfos) {
                     Data key = recordReplicationInfo.getKey();
                     final Data value = recordReplicationInfo.getValue();
                     Record newRecord = recordStore.createRecord(value, -1L, Clock.currentTimeMillis());
                     applyRecordInfo(newRecord, recordReplicationInfo);
                     recordStore.putRecord(key, newRecord);
+
+                    if (!indexes.hasIndex()) {
+                        continue; // no indexes to populate
+                    }
+                    if (indexes.isGlobal()) {
+                        continue; // global indexes are populated during migration finalization
+                    }
+                    if (mapReplicationOperation.getReplicaIndex() != 0) {
+                        continue; // backup partitions have no indexes to populate
+                    }
+                    final Object valueToIndex = Records.getValueOrCachedValue(newRecord, serializationService);
```

Comment:
I copied this line from the existing code, but seems like we don't need to mess here with the caching, so this may be simplified, right?